spring-cloud-gcp-data-datastore/src/main/java/org/springframework/cloud/gcp/data/datastore/core/mapping/DatastoreMappingContext.java

```diff
@@ -67,15 +67,23 @@ public void setApplicationContext(ApplicationContext applicationContext)
 	/**
 	 * Registers in the DatastoreMappingContext that two classes are discriminated from the
 	 * same Datastore Kind.
-	 * @param classA one class.
-	 * @param classB another class.
+	 * @param parentClass the superclass.
+	 * @param subClass the subclass.
 	 */
-	public static void addDiscriminationClassConnection(Class classA, Class classB) {
-		Set<Class> setA = discriminationFamilies.computeIfAbsent(classA, DatastoreMappingContext::createSingletonSet);
-		Set<Class> setB = discriminationFamilies.computeIfAbsent(classB, DatastoreMappingContext::createSingletonSet);
-
-		setA.addAll(setB);
-		discriminationFamilies.put(classB, setA);
+	public static void addDiscriminationClassConnection(Class parentClass, Class subClass) {
+		Set<Class> setParent = discriminationFamilies.computeIfAbsent(parentClass, unused -> new HashSet<>());
+		Set<Class> setSubClass = discriminationFamilies.computeIfAbsent(subClass, unused -> new HashSet<>());
+		setParent.add(subClass);
+
+		setSubClass.forEach(x -> {
+			if (!discriminationFamilies.get(parentClass).contains(x)) {
+				addDiscriminationClassConnection(parentClass, x);
+			}
+		});
+		Class grandParent = parentClass.getSuperclass();
+		if (grandParent != null) {
+			addDiscriminationClassConnection(grandParent, subClass);
+		}
 	}
 
 	private static Set<Class> createSingletonSet(Class aClass) {
```

Comment:
remove `createSingletonSet` if it's unused

Reply:
done